src/plugins/es_ui_shared/static/forms/hook_form_lib/hooks/use_form.ts

```diff
@@ -0,0 +1,198 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { useState, useRef } from 'react';
+import { get } from 'lodash';
+
+import { Form, FormData, FieldConfig, FieldsMap, FormConfig } from '../types';
+import { mapFormFields, flattenObject, unflattenObject, Subject } from '../lib';
+
+const DEFAULT_ERROR_DISPLAY_TIMEOUT = 500;
+
+export const useForm = <T = FormData>(
+  formConfig: FormConfig<T> | undefined = {}
+): { form: Form<T> } => {
+  const {
+    onSubmit,
+    schema,
+    defaultValue = {},
+    serializer = (data: any) => data,
+    deserializer = (data: any) => data,
+    options = { errorDisplayDelay: DEFAULT_ERROR_DISPLAY_TIMEOUT, stripEmptyFields: true },
+  } = formConfig;
+  const defaultValueDeserialized =
+    Object.keys(defaultValue).length === 0 ? defaultValue : deserializer(defaultValue);
+  const [isSubmitted, setSubmitted] = useState(false);
+  const [isSubmitting, setSubmitting] = useState(false);
+  const [isValid, setIsValid] = useState(true);
+  const fieldsRefs = useRef<FieldsMap>({});
+
+  // formData$ is an observable we can subscribe to in order to receive live
+  // update of the raw form data. As an observable it does not trigger any React
+  // render().
+  // The <FormDataProvider> component is the one in charge of reading this observable
+  // and updating its state to trigger the necessary view render.
+  const formData$ = useRef<Subject<T>>(new Subject<T>(flattenObject(defaultValue) as T));
+
+  // -- HELPERS
+  // ----------------------------------
+  const fieldsToArray = () => Object.values(fieldsRefs.current);
+
+  const stripEmptyFields = (fields: FieldsMap): FieldsMap => {
+    if (options.stripEmptyFields) {
+      return Object.entries(fields).reduce(
+        (acc, [key, field]) => {
+          if (field.value !== '') {
```

Comment:
In response to my question about multiple whitespace, you replied:

> Yes multiple whitespace === an empty field.

But in this check `'     ' === ''` will evaluate to false. I just want to double-check that this is what you intended?

Reply:
Good catch 👍